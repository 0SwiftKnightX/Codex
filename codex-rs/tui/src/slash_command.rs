use std::collections::HashMap;

use strum::IntoEnumIterator;
use strum_macros::AsRefStr; // derive macro
use strum_macros::EnumIter;
use strum_macros::EnumString;
use strum_macros::IntoStaticStr;

/// Commands that can be invoked by starting a message with a leading slash.
#[derive(
    Debug, Clone, Copy, PartialEq, Eq, Hash, EnumString, EnumIter, AsRefStr, IntoStaticStr,
)]
#[strum(serialize_all = "kebab-case")]
pub enum SlashCommand {
    New,
    ToggleMouseMode,
    /// Launch the external editor to edit the current prompt draft.
    EditPrompt,
    Quit,
    /// Add a dynamic mount (host path → container path).
    MountAdd,
    /// Remove a dynamic mount by container path.
    MountRemove,
    /// Prompt to run a shell command in the container.
    Shell,
}

impl SlashCommand {
    /// User-visible description shown in the popup.
    pub fn description(self) -> &'static str {
        match self {
            SlashCommand::New => "Start a new chat.",
            SlashCommand::ToggleMouseMode =>
                "Toggle mouse mode (enable for scrolling, disable for text selection)",
            SlashCommand::EditPrompt =>
                "Open external editor to edit the current prompt.",
            SlashCommand::Quit => "Exit the application.",
        SlashCommand::MountAdd => "Add a mount: host path → container path.",
        SlashCommand::MountRemove => "Remove a mount by container path.",
        SlashCommand::Shell => "Run a shell command in the container.",
        }
    }

    /// Command string without the leading '/'. Provided for compatibility with
    /// existing code that expects a method named `command()`.
    pub fn command(self) -> &'static str {
        self.into()
    }
}

/// Return all built-in commands in a HashMap keyed by their command string.
pub fn built_in_slash_commands() -> HashMap<&'static str, SlashCommand> {
    SlashCommand::iter().map(|c| (c.command(), c)).collect()
}
