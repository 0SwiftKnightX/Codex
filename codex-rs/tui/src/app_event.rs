use codex_core::protocol::Event;
use crossterm::event::KeyEvent;

use crate::slash_command::SlashCommand;

#[allow(clippy::large_enum_variant)]
pub(crate) enum AppEvent {
    CodexEvent(Event),

    Redraw,

    KeyEvent(KeyEvent),

    /// Scroll event with a value representing the "scroll delta" as the net
    /// scroll up/down events within a short time window.
    Scroll(i32),

    /// Request to exit the application gracefully.
    ExitRequest,

    /// Forward an `Op` to the Agent. Using an `AppEvent` for this avoids
    /// bubbling channels through layers of widgets.
    CodexOp(codex_core::protocol::Op),

    /// Latest formatted log line emitted by `tracing`.
    LatestLog(String),

    /// Dispatch a recognized slash command from the UI (composer) to the app
    /// layer so it can be handled centrally.
    /// Dispatch a recognized slash command from the UI (composer) to the app
    /// layer so it can be handled centrally (interactive dialog).
    DispatchCommand(SlashCommand),
    /// Inline mount-add DSL: raw argument string (`host=... container=... mode=...`).
    InlineMountAdd(String),
    /// Inline mount-remove DSL: raw argument string (`container=...`).
    InlineMountRemove(String),
    /// Perform mount-add: create symlink and update sandbox policy.
    MountAdd {
        host: std::path::PathBuf,
        container: std::path::PathBuf,
        mode: String,
    },
    /// Perform mount-remove: remove symlink and update sandbox policy.
    MountRemove {
        container: std::path::PathBuf,
    },
}
    /// Notify that the on-disk config.toml has changed and present diff.
    ConfigReloadRequest(String),
    /// Apply the new on-disk config.toml.
    ConfigReloadApply,
    /// Ignore on-disk config.toml changes and continue with old config.
    ConfigReloadIgnore,
