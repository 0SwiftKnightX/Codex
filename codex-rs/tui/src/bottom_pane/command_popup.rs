use ratatui::buffer::Buffer;
use ratatui::layout::Rect;
use ratatui::widgets::WidgetRef;

use super::popup_consts::MAX_POPUP_ROWS;
use super::scroll_state::ScrollState;
use super::selection_popup_common::GenericDisplayRow;
use super::selection_popup_common::render_rows;
use crate::slash_command::SlashCommand;
use crate::slash_command::built_in_slash_commands;
use codex_common::fuzzy_match::fuzzy_match;
use std::fs;
use std::path::PathBuf;

/// A discovered prompt in ~/.codex/prompts.
#[derive(Clone, Debug)]
pub(crate) struct PromptEntry {
    pub name: String,
    pub path: PathBuf,
}

/// A selectable item in the popup: either a built-in command or a user prompt.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub(crate) enum CommandItem {
    Builtin(SlashCommand),
    /// Index into `prompts` vector of `CommandPopup`.
    Prompt(usize),
}

pub(crate) struct CommandPopup {
    command_filter: String,
    builtins: Vec<(&'static str, SlashCommand)>,
    prompts: Vec<PromptEntry>,
    state: ScrollState,
}

impl CommandPopup {
    pub(crate) fn new() -> Self {
        let builtins = built_in_slash_commands();
        let prompts = Self::discover_prompts(&builtins);
        Self {
            command_filter: String::new(),
            builtins,
            prompts,
            state: ScrollState::new(),
        }
    }

    /// If `idx` refers to a prompt item, return its absolute file path.
    pub(crate) fn prompt_path(&self, idx: usize) -> Option<&PathBuf> {
        self.prompts.get(idx).map(|p| &p.path)
    }

    pub(crate) fn prompt_name(&self, idx: usize) -> Option<&str> {
        self.prompts.get(idx).map(|p| p.name.as_str())
    }

    /// Scan ~/.codex/prompts for files and return them as prompt entries.
    fn discover_prompts(builtins: &Vec<(&'static str, SlashCommand)>) -> Vec<PromptEntry> {
        // Build a set of builtin names to avoid collisions.
        let mut builtin_names = std::collections::HashSet::new();
        for (name, _) in builtins.iter() {
            builtin_names.insert(*name);
        }

        let home = std::env::var("HOME").unwrap_or_else(|_| ".".to_string());
        let prompts_dir = PathBuf::from(format!("{home}/.codex/prompts"));
        let mut out = Vec::new();
        if let Ok(entries) = fs::read_dir(&prompts_dir) {
            for entry in entries.flatten() {
                let path = entry.path();
                if !path.is_file() {
                    continue;
                }
                let file_name = path.file_stem().and_then(|s| s.to_str());
                let Some(name) = file_name else {
                    continue;
                };
                // Avoid duplicates with built-ins; prefer built-ins.
                if builtin_names.contains(name) {
                    continue;
                }
                out.push(PromptEntry {
                    name: name.to_string(),
                    path,
                });
            }
            // Sort prompts by name for stable ordering.
            out.sort_by(|a, b| a.name.cmp(&b.name));
        }
        out
    }

    /// Update the filter string based on the current composer text. The text
    /// passed in is expected to start with a leading '/'. Everything after the
    /// *first* '/" on the *first* line becomes the active filter that is used
    /// to narrow down the list of available commands.
    pub(crate) fn on_composer_text_change(&mut self, text: String) {
        let first_line = text.lines().next().unwrap_or("");

        if let Some(stripped) = first_line.strip_prefix('/') {
            // Extract the *first* token (sequence of non-whitespace
            // characters) after the slash so that `/clear something` still
            // shows the help for `/clear`.
            let token = stripped.trim_start();
            let cmd_token = token.split_whitespace().next().unwrap_or("");

            // Update the filter keeping the original case (commands are all
            // lower-case for now but this may change in the future).
            self.command_filter = cmd_token.to_string();
        } else {
            // The composer no longer starts with '/'. Reset the filter so the
            // popup shows the *full* command list if it is still displayed
            // for some reason.
            self.command_filter.clear();
        }

        // Reset or clamp selected index based on new filtered list.
        let matches_len = self.filtered_items().len();
        self.state.clamp_selection(matches_len);
        self.state
            .ensure_visible(matches_len, MAX_POPUP_ROWS.min(matches_len));
    }

    /// Determine the preferred height of the popup. This is the number of
    /// rows required to show at most MAX_POPUP_ROWS commands.
    pub(crate) fn calculate_required_height(&self) -> u16 {
        self.filtered_items().len().clamp(1, MAX_POPUP_ROWS) as u16
    }

    /// Compute fuzzy-filtered matches paired with optional highlight indices and score.
    /// Sorted by ascending score, then by command name for stability.
    fn filtered(&self) -> Vec<(CommandItem, Option<Vec<usize>>, i32)> {
        let filter = self.command_filter.trim();
        let mut out: Vec<(CommandItem, Option<Vec<usize>>, i32)> = Vec::new();
        if filter.is_empty() {
            // Built-ins first, in presentation order.
            for (_, cmd) in self.builtins.iter() {
                out.push((CommandItem::Builtin(*cmd), None, 0));
            }
            // Then prompts, already sorted by name.
            for idx in 0..self.prompts.len() {
                out.push((CommandItem::Prompt(idx), None, 0));
            }
            return out;
        }

        for (_, cmd) in self.builtins.iter() {
            if let Some((indices, score)) = fuzzy_match(cmd.command(), filter) {
                out.push((CommandItem::Builtin(*cmd), Some(indices), score));
            }
        }
        for (idx, p) in self.prompts.iter().enumerate() {
            if let Some((indices, score)) = fuzzy_match(&p.name, filter) {
                out.push((CommandItem::Prompt(idx), Some(indices), score));
            }
        }
        // When filtering, sort by ascending score and then by name for stability.
        out.sort_by(|a, b| {
            let an = match a.0 {
                CommandItem::Builtin(c) => c.command(),
                CommandItem::Prompt(i) => &self.prompts[i].name,
            };
            let bn = match b.0 {
                CommandItem::Builtin(c) => c.command(),
                CommandItem::Prompt(i) => &self.prompts[i].name,
            };
            a.2.cmp(&b.2).then_with(|| an.cmp(bn))
        });
        out
    }

    fn filtered_items(&self) -> Vec<CommandItem> {
        self.filtered().into_iter().map(|(c, _, _)| c).collect()
    }

    /// Move the selection cursor one step up.
    pub(crate) fn move_up(&mut self) {
        let len = self.filtered_items().len();
        self.state.move_up_wrap(len);
        self.state.ensure_visible(len, MAX_POPUP_ROWS.min(len));
    }

    /// Move the selection cursor one step down.
    pub(crate) fn move_down(&mut self) {
        let matches_len = self.filtered_items().len();
        self.state.move_down_wrap(matches_len);
        self.state
            .ensure_visible(matches_len, MAX_POPUP_ROWS.min(matches_len));
    }

    /// Return currently selected command, if any.
    pub(crate) fn selected_item(&self) -> Option<CommandItem> {
        let matches = self.filtered_items();
        self.state
            .selected_idx
            .and_then(|idx| matches.get(idx).copied())
    }
}

impl WidgetRef for CommandPopup {
    fn render_ref(&self, area: Rect, buf: &mut Buffer) {
        let matches = self.filtered();
        let rows_all: Vec<GenericDisplayRow> = if matches.is_empty() {
            Vec::new()
        } else {
            matches
                .into_iter()
                .map(|(item, indices, _)| match item {
                    CommandItem::Builtin(cmd) => GenericDisplayRow {
                        name: format!("/{}", cmd.command()),
                        match_indices: indices.map(|v| v.into_iter().map(|i| i + 1).collect()),
                        is_current: false,
                        description: Some(cmd.description().to_string()),
                    },
                    CommandItem::Prompt(i) => GenericDisplayRow {
                        name: format!("/{}", self.prompts[i].name),
                        match_indices: indices.map(|v| v.into_iter().map(|i| i + 1).collect()),
                        is_current: false,
                        description: Some("send saved prompt".to_string()),
                    },
                })
                .collect()
        };
        render_rows(area, buf, &rows_all, &self.state, MAX_POPUP_ROWS, false);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[test]
    fn filter_includes_init_when_typing_prefix() {
        let mut popup = CommandPopup::new();
        // Simulate the composer line starting with '/in' so the popup filters
        // matching commands by prefix.
        popup.on_composer_text_change("/in".to_string());

        // Access the filtered list via the selected command and ensure that
        // one of the matches is the new "init" command.
        let matches = popup.filtered_items();
        let has_init = matches.iter().any(|item| match item {
            CommandItem::Builtin(cmd) => cmd.command() == "init",
            CommandItem::Prompt(_) => false,
        });
        assert!(
            has_init,
            "expected '/init' to appear among filtered commands"
        );
    }

    #[test]
    fn selecting_init_by_exact_match() {
        let mut popup = CommandPopup::new();
        popup.on_composer_text_change("/init".to_string());

        // When an exact match exists, the selected command should be that
        // command by default.
        let selected = popup.selected_item();
        match selected {
            Some(CommandItem::Builtin(cmd)) => assert_eq!(cmd.command(), "init"),
            Some(CommandItem::Prompt(_)) => panic!("unexpected prompt selected for '/init'"),
            None => panic!("expected a selected command for exact match"),
        }
    }

    #[test]
    fn prompt_discovery_lists_custom_prompts() {
        let tmp = tempdir().expect("create TempDir");
        let home = tmp.path();
        let prompts_dir = home.join(".codex").join("prompts");
        std::fs::create_dir_all(&prompts_dir).expect("mkdir -p ~/.codex/prompts");
        std::fs::write(prompts_dir.join("foo"), b"hello from foo").unwrap();
        std::fs::write(prompts_dir.join("bar"), b"hello from bar").unwrap();

        // Point HOME to the temp dir so discovery uses our fixtures.
        unsafe { std::env::set_var("HOME", home) };

        let popup = CommandPopup::new();
        let items = popup.filtered_items();
        let mut prompt_names: Vec<String> = items
            .into_iter()
            .filter_map(|it| match it {
                CommandItem::Prompt(i) => popup.prompt_name(i).map(|s| s.to_string()),
                _ => None,
            })
            .collect();
        prompt_names.sort();
        assert_eq!(prompt_names, vec!["bar".to_string(), "foo".to_string()]);
    }

    #[test]
    fn prompt_name_collision_with_builtin_is_ignored() {
        let tmp = tempdir().expect("create TempDir");
        let home = tmp.path();
        let prompts_dir = home.join(".codex").join("prompts");
        std::fs::create_dir_all(&prompts_dir).expect("mkdir -p ~/.codex/prompts");
        // Create a prompt with the same name as a builtin command (e.g. "init").
        std::fs::write(prompts_dir.join("init"), b"should be ignored").unwrap();

        unsafe { std::env::set_var("HOME", home) };

        let popup = CommandPopup::new();
        let items = popup.filtered_items();
        let has_collision_prompt = items.into_iter().any(|it| match it {
            CommandItem::Prompt(i) => popup.prompt_name(i) == Some("init"),
            _ => false,
        });
        assert!(
            !has_collision_prompt,
            "prompt with builtin name should be ignored"
        );
    }
}
