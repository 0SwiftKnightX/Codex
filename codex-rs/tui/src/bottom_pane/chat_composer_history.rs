use std::collections::HashMap;
use std::sync::mpsc::Sender;

use tui_textarea::CursorMove;
use tui_textarea::TextArea;

use crate::app_event::AppEvent;
use codex_core::protocol::Op;

/// State machine that manages shell-style history navigation (Up/Down) inside
/// the chat composer. This struct is intentionally decoupled from the
/// rendering widget so the logic remains isolated and easier to test.
pub(crate) struct ChatComposerHistory {
    /// Identifier of the history log as reported by `SessionConfiguredEvent`.
    history_log_id: Option<u64>,
    /// Number of entries already present in the persistent cross-session
    /// history file when the session started.
    history_entry_count: usize,

    /// Messages submitted by the user *during this UI session* (newest at END).
    local_history: Vec<String>,

    /// Cache of persistent history entries fetched on-demand.
    fetched_history: HashMap<usize, String>,

    /// Current cursor within the combined (persistent + local) history. `None`
    /// indicates the user is *not* currently browsing history.
    history_cursor: Option<isize>,

    /// The text that was last inserted into the composer as a result of
    /// history navigation. Used to decide if further Up/Down presses should be
    /// treated as navigation versus normal cursor movement.
    last_history_text: Option<String>,
}

impl ChatComposerHistory {
    pub fn new() -> Self {
        Self {
            history_log_id: None,
            history_entry_count: 0,
            local_history: Vec::new(),
            fetched_history: HashMap::new(),
            history_cursor: None,
            last_history_text: None,
        }
    }

    /// Update metadata when a new session is configured.
    pub fn set_metadata(&mut self, log_id: u64, entry_count: usize) {
        self.history_log_id = Some(log_id);
        self.history_entry_count = entry_count;
        self.fetched_history.clear();
        self.local_history.clear();
        self.history_cursor = None;
        self.last_history_text = None;
    }

    /// Record a message submitted by the user in the current session so it can
    /// be recalled later.
    pub fn record_local_submission(&mut self, text: &str) {
        if !text.is_empty() {
            self.local_history.push(text.to_string());
            self.history_cursor = None;
            self.last_history_text = None;
        }
    }

    /// Should Up/Down key presses be interpreted as history navigation given
    /// the current content and cursor position of `textarea`?
    pub fn should_handle_navigation(&self, textarea: &TextArea) -> bool {
        if self.history_entry_count == 0 && self.local_history.is_empty() {
            return false;
        }

        let lines = textarea.lines();
        if lines.len() == 1 && lines[0].is_empty() {
            return true;
        }

        // Textarea is not empty – only navigate when cursor is at start and
        // text matches last recalled history entry so regular editing is not
        // hijacked.
        let (row, col) = textarea.cursor();
        if row != 0 || col != 0 {
            return false;
        }

        matches!(&self.last_history_text, Some(prev) if prev == &lines.join("\n"))
    }

    /// Handle <Up>. Returns true when the key was consumed and the caller
    /// should request a redraw.
    pub fn navigate_up(
        &mut self,
        textarea: &mut TextArea,
        app_event_tx: &Sender<AppEvent>,
    ) -> bool {
        let total_entries = self.history_entry_count + self.local_history.len();
        if total_entries == 0 {
            return false;
        }

        let next_idx = match self.history_cursor {
            None => (total_entries as isize) - 1,
            Some(0) => return true, // already at oldest
            Some(idx) => idx - 1,
        };

        self.history_cursor = Some(next_idx);
        self.populate_history_at_index(next_idx as usize, textarea, app_event_tx);
        true
    }

    /// Handle <Down>.
    pub fn navigate_down(
        &mut self,
        textarea: &mut TextArea,
        app_event_tx: &Sender<AppEvent>,
    ) -> bool {
        let total_entries = self.history_entry_count + self.local_history.len();
        if total_entries == 0 {
            return false;
        }

        let next_idx_opt = match self.history_cursor {
            None => return false, // not browsing
            Some(idx) if (idx as usize) + 1 >= total_entries => None,
            Some(idx) => Some(idx + 1),
        };

        match next_idx_opt {
            Some(idx) => {
                self.history_cursor = Some(idx);
                self.populate_history_at_index(idx as usize, textarea, app_event_tx);
            }
            None => {
                // Past newest – clear and exit browsing mode.
                self.history_cursor = None;
                self.last_history_text = None;
                Self::replace_textarea_content(textarea, "");
            }
        }
        true
    }

    /// Integrate a GetHistoryEntryResponse event.
    pub fn on_entry_response(
        &mut self,
        log_id: u64,
        offset: usize,
        entry: Option<String>,
        textarea: &mut TextArea,
    ) {
        if self.history_log_id != Some(log_id) {
            return;
        }
        let Some(text) = entry else { return };
        self.fetched_history.insert(offset, text.clone());

        if self.history_cursor == Some(offset as isize) {
            Self::replace_textarea_content(textarea, &text);
        }
    }

    // ---------------------------------------------------------------------
    // Internal helpers
    // ---------------------------------------------------------------------

    fn populate_history_at_index(
        &mut self,
        global_idx: usize,
        textarea: &mut TextArea,
        app_event_tx: &Sender<AppEvent>,
    ) {
        if global_idx >= self.history_entry_count {
            // Local entry.
            if let Some(text) = self
                .local_history
                .get(global_idx - self.history_entry_count)
            {
                Self::replace_textarea_content(textarea, text);
            }
        } else if let Some(text) = self.fetched_history.get(&global_idx) {
            Self::replace_textarea_content(textarea, text);
        } else if let Some(log_id) = self.history_log_id {
            let op = Op::GetHistoryEntryRequest {
                offset: global_idx,
                log_id,
            };
            let _ = app_event_tx.send(AppEvent::CodexOp(op));
        }
    }

    fn replace_textarea_content(textarea: &mut TextArea, text: &str) {
        textarea.select_all();
        textarea.cut();
        let _ = textarea.insert_str(text);
        textarea.move_cursor(CursorMove::Jump(0, 0));
    }
}
