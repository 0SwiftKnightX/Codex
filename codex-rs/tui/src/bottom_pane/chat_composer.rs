use crossterm::event::KeyEvent;
use ratatui::buffer::Buffer;
use ratatui::layout::Alignment;
use ratatui::layout::Rect;
use ratatui::style::{Color, Style, Stylize};
use ratatui::text::Line;
use ratatui::widgets::BorderType;
use ratatui::widgets::Borders;
use ratatui::widgets::Widget;
use ratatui::widgets::WidgetRef;
use tui_textarea::Input;
use tui_textarea::Key;
use tui_textarea::TextArea;

use super::chat_composer_history::ChatComposerHistory;
use super::command_popup::CommandPopup;

use crate::app_event::AppEvent;
use crate::app_event_sender::AppEventSender;
use crate::slash_command::SlashCommand;

/// Minimum number of visible text rows inside the textarea.
const MIN_TEXTAREA_ROWS: usize = 1;
/// Rows consumed by the border.
const BORDER_LINES: u16 = 2;

/// Result returned when the user interacts with the text area.
pub enum InputResult {
    Submitted(String),
    None,
}

pub(crate) struct ChatComposer<'a> {
    textarea: TextArea<'a>,
    command_popup: Option<CommandPopup>,
    app_event_tx: AppEventSender,
    history: ChatComposerHistory,
    /// Maximum number of visible lines in the chat input composer.
    max_rows: usize,
    /// Last computed context-left percentage
    context_left_percent: f64,
}

impl ChatComposer<'_> {
    pub fn new(has_input_focus: bool, app_event_tx: AppEventSender, max_rows: usize) -> Self {
        let mut textarea = TextArea::default();
        textarea.set_placeholder_text("send a message");
        textarea.set_cursor_line_style(ratatui::style::Style::default());

        let mut this = Self {
            textarea,
            command_popup: None,
            app_event_tx,
            history: ChatComposerHistory::new(),
            max_rows,
            context_left_percent: 100.0,
        };
        this.update_border(has_input_focus);
        this
    }

    /// Record the history metadata advertised by `SessionConfiguredEvent` so
    /// that the composer can navigate cross-session history.
    pub(crate) fn set_history_metadata(&mut self, log_id: u64, entry_count: usize) {
        self.history.set_metadata(log_id, entry_count);
    }

    /// Integrate an asynchronous response to an on-demand history lookup. If
    /// the entry is present and the offset matches the current cursor we
    /// immediately populate the textarea.
    pub(crate) fn on_history_entry_response(
        &mut self,
        log_id: u64,
        offset: usize,
        entry: Option<String>,
    ) -> bool {
        self.history
            .on_entry_response(log_id, offset, entry, &mut self.textarea)
    }

    pub fn set_input_focus(&mut self, has_focus: bool) {
        self.update_border(has_focus);
    }

    /// Update the context-left percentage for display.
    pub fn set_context_left(&mut self, pct: f64) {
        self.context_left_percent = pct;
    }

    /// Handle a key event coming from the main UI.
    pub fn handle_key_event(&mut self, key_event: KeyEvent) -> (InputResult, bool) {
        let result = match self.command_popup {
            Some(_) => self.handle_key_event_with_popup(key_event),
            None => self.handle_key_event_without_popup(key_event),
        };

        // Update (or hide/show) popup after processing the key.
        self.sync_command_popup();

        result
    }

    /// Handle key event when the slash-command popup is visible.
    fn handle_key_event_with_popup(&mut self, key_event: KeyEvent) -> (InputResult, bool) {
        let Some(popup) = self.command_popup.as_mut() else {
            tracing::error!("handle_key_event_with_popup called without an active popup");
            return (InputResult::None, false);
        };

        match key_event.into() {
            Input { key: Key::Up, .. } => {
                popup.move_up();
                (InputResult::None, true)
            }
            Input { key: Key::Down, .. } => {
                popup.move_down();
                (InputResult::None, true)
            }
            Input { key: Key::Tab, .. } => {
                if let Some(cmd) = popup.selected_command() {
                    let first_line = self
                        .textarea
                        .lines()
                        .first()
                        .map(|s| s.as_str())
                        .unwrap_or("");

                    let starts_with_cmd = first_line
                        .trim_start()
                        .starts_with(&format!("/{}", cmd.command()));

                    if !starts_with_cmd {
                        self.textarea.select_all();
                        self.textarea.cut();
                        let _ = self.textarea.insert_str(format!("/{} ", cmd.command()));
                    }
                }
                (InputResult::None, true)
            }
            Input { key: Key::Enter, shift: false, alt: false, ctrl: false } => {
                if let Some(cmd) = popup.selected_command() {
                    self.textarea.select_all();
                    self.textarea.cut();
                    self.command_popup = None;
                    return (InputResult::None, true);
                }
                // Fallback to default newline handling if no command selected.
                self.handle_key_event_without_popup(key_event)
            }
            input => self.handle_input_basic(input),
        }
    }

    /// Handle key event when no popup is visible.
    fn handle_key_event_without_popup(&mut self, key_event: KeyEvent) -> (InputResult, bool) {
        let input: Input = key_event.into();
        match input {
            // -------------------------------------------------------------
            // History navigation (Up / Down) â€“ only when the composer is not
            // empty or when the cursor is at the correct position, to avoid
            // interfering with normal cursor movement.
            // -------------------------------------------------------------
            Input { key: Key::Up, .. } => {
                if self.history.should_handle_navigation(&self.textarea) {
                    let consumed = self
                        .history
                        .navigate_up(&mut self.textarea, &self.app_event_tx);
                    if consumed {
                        return (InputResult::None, true);
                    }
                }
                self.handle_input_basic(input)
            }
            Input { key: Key::Down, .. } => {
                if self.history.should_handle_navigation(&self.textarea) {
                    let consumed = self
                        .history
                        .navigate_down(&mut self.textarea, &self.app_event_tx);
                    if consumed {
                        return (InputResult::None, true);
                    }
                }
                self.handle_input_basic(input)
            }
            Input {
                key: Key::Enter,
                shift: false,
                alt: false,
                ctrl: false,
            } => {
                let text = self.textarea.lines().join("\n");
                self.textarea.select_all();
                self.textarea.cut();

                if text.is_empty() {
                    (InputResult::None, true)
                } else {
                    self.history.record_local_submission(&text);
                    (InputResult::Submitted(text), true)
                }
            }
            Input { key: Key::Enter, .. }
            | Input { key: Key::Char('j'), ctrl: true, alt: false, shift: false } => {
                self.textarea.insert_newline();
                (InputResult::None, true)
            }
            Input { key: Key::Char('e'), ctrl: true, alt: false, shift: false } => {
                // Launch external editor for prompt drafting
                self.open_external_editor();
                (InputResult::None, true)
            }
            input => self.handle_input_basic(input),
        }
    }

    /// Handle generic Input events that modify the textarea content.
    fn handle_input_basic(&mut self, input: Input) -> (InputResult, bool) {
        self.textarea.input(input);
        (InputResult::None, true)
    }

    /// Launch an external editor on a temporary file pre-populated with the current draft,
    /// then reload the edited contents back into the textarea on exit.
    pub fn open_external_editor(&mut self) {
        use std::io::Write;
        use std::process::Command;
        // Dump current draft to a temp file
        let content = self.textarea.lines().join("\n");
        let mut tmp = match tempfile::NamedTempFile::new() {
            Ok(f) => f,
            Err(e) => {
                tracing::error!("failed to create temp file for editor: {e}");
                return;
            }
        };
        if let Err(e) = write!(tmp, "{}", content) {
            tracing::error!("failed to write to temp file for editor: {e}");
            return;
        }
        let path = tmp.path();
        // Determine editor: VISUAL > EDITOR > nvim
        let editor = std::env::var("VISUAL").or_else(|_| std::env::var("EDITOR")).unwrap_or_else(|_| "nvim".into());
        // Launch editor and wait for exit
        if let Err(e) = Command::new(editor).arg(path).status() {
            tracing::error!("failed to launch editor: {e}");
            return;
        }
        // Read back edited contents (fall back to original on error)
        let new_text = std::fs::read_to_string(path).unwrap_or(content);
        // Replace textarea contents
        self.textarea.select_all();
        self.textarea.cut();
        let _ = self.textarea.insert_str(new_text);
    }

    /// Synchronize `self.command_popup` with the current text in the
    /// textarea. This must be called after every modification that can change
    /// the text so the popup is shown/updated/hidden as appropriate.
    fn sync_command_popup(&mut self) {
        // Inspect only the first line to decide whether to show the popup. In
        // the common case (no leading slash) we avoid copying the entire
        // textarea contents.
        let first_line = self
            .textarea
            .lines()
            .first()
            .map(|s| s.as_str())
            .unwrap_or("");

        if first_line.starts_with('/') {
            // Create popup lazily when the user starts a slash command.
            let popup = self.command_popup.get_or_insert_with(CommandPopup::new);

            // Forward *only* the first line since `CommandPopup` only needs
            // the command token.
            popup.on_composer_text_change(first_line.to_string());
        } else if self.command_popup.is_some() {
            // Remove popup when '/' is no longer the first character.
            self.command_popup = None;
        }
    }

    pub fn calculate_required_height(&self, area: &Rect) -> u16 {
        let rows = self
            .textarea
            .lines()
            .len()
            .max(MIN_TEXTAREA_ROWS)
            .min(self.max_rows);
        let num_popup_rows = if let Some(popup) = &self.command_popup {
            popup.calculate_required_height(area)
        } else {
            0
        };

        // Include an extra row for the context-left indicator when not in popup mode
        let context_row = if self.command_popup.is_none() { 1 } else { 0 };
        rows as u16 + BORDER_LINES + num_popup_rows + context_row
    }

    fn update_border(&mut self, has_focus: bool) {
        struct BlockState {
            right_title: Line<'static>,
            border_style: Style,
        }

        let bs = if has_focus {
            BlockState {
                right_title: Line::from("Enter to send | Ctrl+D to quit | Ctrl+J for newline")
                    .alignment(Alignment::Right),
                border_style: Style::default(),
            }
        } else {
            BlockState {
                right_title: Line::from(""),
                border_style: Style::default().dim(),
            }
        };

        self.textarea.set_block(
            ratatui::widgets::Block::default()
                .title_bottom(bs.right_title)
                .borders(Borders::ALL)
                .border_type(BorderType::Rounded)
                .border_style(bs.border_style),
        );
    }

    pub(crate) fn is_command_popup_visible(&self) -> bool {
        self.command_popup.is_some()
    }
}

impl WidgetRef for &ChatComposer<'_> {
    fn render_ref(&self, area: Rect, buf: &mut Buffer) {
        if let Some(popup) = &self.command_popup {
            let popup_height = popup.calculate_required_height(&area);

            // Split the provided rect so that the popup is rendered at the
            // *top* and the textarea occupies the remaining space below.
            let popup_rect = Rect {
                x: area.x,
                y: area.y,
                width: area.width,
                height: popup_height.min(area.height),
            };

            let textarea_rect = Rect {
                x: area.x,
                y: area.y + popup_rect.height,
                width: area.width,
                height: area.height.saturating_sub(popup_rect.height),
            };

            popup.render(popup_rect, buf);
            self.textarea.render(textarea_rect, buf);
        } else {
            self.textarea.render(area, buf);
        }
        // Render context-left indicator when not displaying a popup
        if self.command_popup.is_none() {
            let pct = self.context_left_percent.round();
            let text = format!("{:.0}% context left", pct);
            let color = if pct > 40.0 {
                Color::Green
            } else if pct > 25.0 {
                Color::Yellow
            } else {
                Color::Red
            };
            buf.set_string(area.x + 1, area.y + area.height - 1, text, Style::default().fg(color));
        }
    }
}
