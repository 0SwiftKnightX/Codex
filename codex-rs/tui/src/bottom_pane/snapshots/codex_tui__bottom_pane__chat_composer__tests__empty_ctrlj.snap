---
source: tui/src/bottom_pane/chat_composer.rs
expression: terminal_ctrlj.backend()
---
"▌ ...                                                                                               "
"▌                                                                                                   "
"▌                                                                                                   "
"▌                                                                                                   "
"▌                                                                                                   "
"▌                                                                                                   "
"▌                                                                                                   "
"▌                                                                                                   "
"▌                                                                                                   "
" ⏎ send   Ctrl+J newline   Ctrl+C quit                                                              "
